plutarch-test/goldens/unit.pmatch.bench.golden

```diff
@@ -0,0 +1 @@
+0 {"exBudgetCPU":29873,"exBudgetMemory":200,"scriptSizeBytes":6}
```

Comment:
No `uplc` results for this?

Reply:
@L-as Some tests specifically generate _only_ bench goldens (_or_ only uplc goldens), rather than both, as controlled by this type:

https://github.com/Plutonomicon/plutarch/blob/1ebeff5b2dcf5205c753a55b6efe430b2c7f5b00/plutarch-test/src/Plutarch/Test.hs#L135-L149

That's what is happening here, because of `golden Bench p` (instead of `golden All p`). Should we get rid of `Bench` and *always* generate printTerm regardless of the program? Likewise, are there cases where avoiding benchmarks make sense? It is used in a handful of cases, like:

```
plutarch-test/src/Plutarch/PLamSpec.hs
12:      golden PrintTerm $ plam (\x -> x)
14:      golden PrintTerm $ plam (\_ y -> y)
16:      golden PrintTerm $ plam (\x _ -> plet x $ \_ -> perror)
```

I figured we don't need to benchmark lambdas, inasmuch as the body of a lambda won't be evaluated anyway. But may be it is worth benchmarking the script size anyway.

Reply:
Here's an example of something generating bench goldens, but no printTerm golden:

```haskell
    describe "phexByteStr" $ do
      let a :: [String] = ["42", "ab", "df", "c9"]
          p = pconstant @PByteString (BS.pack $ map readByte a) #== phexByteStr (concat a)
      golden Bench p
      it "relation" $ passert p
```

At least the eval uplc would be useless here, as it would be: `(program 1.0.0 True)`. The non-eval UPLC would be: `(program 1.0.0 ((\i0 -> equalsByteString i1 i1) #42abdfc9))` - however why should it be under "phexByteStr" tree (see above), rather a tree for equalsByteString?

Reply:
Here, we can do benchmark golden, because doing that will result in error:

```haskell
    describe "pmatch" $ do
      let p = pmatch (integerList [1, 3, 1]) $ \_ -> perror
      golden PrintTerm p
```

Changing `PrintTerm` -> `All` will result in:

<img width="472" alt="image" src="https://user-images.githubusercontent.com/3998/154169936-860b64d3-c75d-451e-8da8-1e278ca619ef.png">

Reply:
Another example, where we can't really benchmark a scriptcontext constant:

```haskell
golden PrintTerm ctx
```
<img width="790" alt="image" src="https://user-images.githubusercontent.com/3998/154170970-422ec836-d024-484e-bc5f-07980fcabbf7.png">

Reply:
Anyway, uplc goldens added in [a001bde](https://github.com/Plutonomicon/plutarch/pull/276/commits/a001bdedb021b4314f3bebad570538ea39cfe4c0)

Reply:
> At least the post-eval uplc would be useless here, as it would be: (program 1.0.0 True). The pre-eval UPLC would be: (program 1.0.0 ((\i0 -> equalsByteString i1 i1) #42abdfc9)) - however why should it be under "phexByteStr" tree (see above), rather a tree for equalsByteString?

I don't agree with this. the post-eval UPLC is still very important, because if it starts evaluating into something else, we have a bug! 9/10 times changes to the golden files are accidental and means we have a bug.

I don't see why we wouldn't support only `All`. Terms that err can still be benchmarked. It just means that the benchmarking result should include only the script size, and possibly still CPU and memory cost if that information isn't lost.

Wrt. the `ctx` case, I'm not sure why that's erring in the first place, but I think that we should still benchmark this too.

You could potentially include the error message *in* the benchmarking golden file.